### PR TITLE
Support for user new password validation

### PR DESCRIPTION
**Pull request by ArturKlajnerok**
This will fix problems with reset password functionality.

/cc @Wikia/services-team

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -181,6 +181,13 @@ func (user *User) IsValidPassword(password string) bool {
 		return false
 	}
 
+	if len(user.NewPassword) > 3 {
+		salt, passHash := ExtractHashAndSalt(user.NewPassword)
+		if passHash == HashPassword(password, salt) {
```
Comment (frankfarmer) [inline models/user.go]: This comparison is vulnerable to timing attacks, right?

http://verboselogging.com/2012/08/20/a-timing-attack-in-action

File: models/user.go [hunk 2]
```diff
@@ -181,6 +181,13 @@ func (user *User) IsValidPassword(password string) bool {
...
+			return true
```
Comment (nmonterroso) [inline models/user.go]: should this return `false` if the extracted hash doesn't match the hash of the user input?

File: models/user.go (hunk 1 above)
Comment (drsnyder) [inline models/user.go]: Below (not shown) the return value is `passHash == HashPassword(password, salt)`. For consistency, should the same be done here?

Comment (ArturKlajnerok) [inline models/user.go]: Thanks @frankfarmer I will look into that for general authentication security :panda_face:

File: models/user.go (hunk 2 above)
Comment (ArturKlajnerok) [inline models/user.go]: @drsnyder & @nmonterroso The reason is to check if password matches `user_newpassword` and if not proceed and make an match with `user_password`.

Comment (nmonterroso) [inline models/user.go]: ah ok, so even after changing your password you can login with your old password

**Comment by nmonterroso** [comment]
LGTM

**Comment by drsnyder** [comment]
From the description, it's not clear the problem you are trying to solve. Can you say more about what the problem was? It's also not clear from the function what constitutes a valid password. 

I think there should also be some unit tests for the function and the added check. Without tests, future us won't know what we can or can't change.

**Comment by ArturKlajnerok** [comment]
@drsnyder I've tried address all your feedback in : https://github.com/Wikia/helios/pull/26

**Pull request merged**